Added Terraform to existing style guide (#101)
GoogleCloudPlatform/samples-style-guide: content/en/_index.md

1. **Clarity**: The code's purpose and rationale is clear to the reader
2. **Idiomaticity**: The code should be written using idiomatic,

>    community-driven coding practices, except where these practices

Reviewer: +1 
This is an important exception for Terraform code samples

Author: To differentiate from our Terraform blueprints (https://cloud.google.com/docs/terraform/blueprints/terraform-blueprints), Terraform resource samples at go/tf-samples aren't allowing abstraction via modules, global variables, loops, etc. These are community-driven coding practices for Terraform, but we don't allow them.

Author: Captured in b/271024053 for further discussion, so closing here.

}
// [END product_example]
{{< /tab >}}

{{< tab header="Terraform" >}}
// Terraform doesn't have `import` statements. However, Terraform
// has the concept of importing resources to bring them under
// Terraform management. For more information, see

> // [Import your Google Cloud resources into Terraform state](https://cloud.google.com/docs/terraform/resource-management/import).

Reviewer: Hyperlink doesn't render in the tab view

Author: Removed content

> // [Import your Google Cloud resources into Terraform state](https://cloud.google.com/docs/terraform/resource-management/import).

Author: done

> // Region tags should start after the copyright and license agreement.

Author: Removed from this PR.

# Copyright 2023 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

// [START product_example]

resource "example_resource" "resource_reference" {

>   ...

Reviewer: replace `...` with `# Snippet arguments` or `# Resource arguments`

Author: Done

}
// [END product_example]
{{< /tab >}}
{{< /tabpane >}}

### One sample per file {#one-per-file}

A single file should only include one sample. This keeps the end-to-end
code relevant to the reader's current learning need.

For languages that have import statements, this shows the minimal
set of dependencies via import statements and helps reviewers validate the

> [imports](#imports) guideline.

Author: Tracking in b/271014774, so closing here.

 * before running the code snippet.
 */
{{< /tab >}}
{{< tab header="Terraform" >}}

> // [START product_example]

Reviewer: Other langauge tags do not include a region start or end tag

Author: fixed

> #   Triggers DAGs Using Cloud Function and Pub/Sub Messages with Terraform

Reviewer: Other language tabs include an example linking to prerequisite documentation such as

```
 * See https://cloud.google.com/compute/docs/quickstart-client-libraries
 * before running the code snippet.
```

Author: Fixed

{{< tab header="PHP" >}}
// This is an example snippet for showing best practices.
function example_snippet(string $projectId, string $filePath): void
{{< /tab >}}

> {{< tab header="Terraform" >}}

Author: Tracking in b/271016879, so closing here.

#  These are example snippets for showing best practices.
#
# Don't include the `project` argument in resources.
# The guidance at https://cloud.google.com/docs/terraform/basic-commands
# shows how to use an `export` statement to set the Google Cloud project

> # where you want to apply the Terraform configuration.

Author: I don't understand the question.
Both what?
I think environment variable is universally recommended. Maybe we should be explicit about that? And maybe even include the instruction in every sample?

> # where you want to apply the Terraform configuration.

Author: Captured in b/271014775, so closing here.

#
# If you run the `export GOOGLE_CLOUD_PROJECT` command, most resources
# can infer the project ID.

#  Creates a VPC network
resource "google_compute_network" "default" {
  name                    = "my-network"
  auto_create_subnetworks = false

> }

Author: done

#
# Some resources, such as `project_iam_*`, cannot infer the project ID.
# As a workaround, use the `data "google_project"` data source.
#
# Creates an IAM policy

> data "google_project" "project" {

Author: Done

> data "google_project" "project" {

Author: done

}

data "google_iam_policy" "sql_iam_policy" {
  binding {
    role = "roles/cloudsql.client"
    members = [
      "serviceAccount:${google_project_service_identity.gcp_sa_cloud_sql.email}",
    ]
    condition {
      expression  = "resource.name == 'projects/${data.google_project.project.id}/instances/${google_sql_database_instance.default.name}' && resource.type == 'sqladmin.googleapis.com/Instance'"
      title       = "created"
      description = "Cloud SQL instance creation"
    }
  }
}

resource "google_project_iam_policy" "project" {
  project     = data.google_project.project.id
  policy_data = data.google_iam_policy.sql_iam_policy.policy_data

> } 

Author: done

// This is an example snippet for showing best practices.
function example_snippet(string $projectId, string $filePath): void
{{< /tab >}}
{{< tab header="Terraform" >}}

> // [START product_example]

Author: fixed

> #   Creates a VPC network

Reviewer: Replace with `# This is an example snippet for showing best practices` to match others.

Would this be a good place to say whether we should include the `project` argument or not? I know some resources will just use the provider project, but others will explicitly look for the `project` argument.

If we do want to say something specifically about the project argument, we can mention that the `project` arg should use the `google_project` data object and include an example of that

    );
}
{{< /tab >}}
{{< tab header="Terraform" >}}

> Terraform doesn't have custom, user-defined methods or functions.

Reviewer: Make this a comment. Markdown url's don't render in tabs, so just include the URL

Author: done

For more information, see [Built-in Functions](https://developer.hashicorp.com/terraform/language/functions).

> {{< /tab >}}

Author: done

    }
}
{{< /tab >}}

> {{< tab header="Terraform" >}}

Reviewer 2: +1. Guideline examples are not meant to be a complete sample app. Examples should be as brief as possible to illustrate the guideline, as opposed to demonstrating a canonical example of the guideline in action.

#StyleGuideFeedback

> {{< tab header="Terraform" >}}

Author: Done

Define any local variables at the beginning of the sample.
Include any output statements at the end of the sample.

> 

Reviewer: throw into a comment block

Author: done

];
$storage = new StorageClient($config);
{{< /tab >}}
{{< tab header="Terraform" >}}
# Use application default credentials when running locally.
# When developers are locally iterating on Terraform configuration,
# they should authenticate by running `gcloud auth application-default login`
# to generate application default credentials. Don't download service account keys,
# because downloaded keys are harder to manage and secure.

> {{< /tab >}}

Author: ACK

];
$storage = new StorageClient($config);
{{< /tab >}}
{{< tab header="Terraform" >}}
Use application default credentials when running locally.
When developers are locally iterating on Terraform configuration,
they should authenticate by running `gcloud auth application-default login`
to generate application default credentials. Don't  download service account keys,

> because downloaded  keys are harder to manage and secure.

Reviewer: Throw into a comment block

Author: done

$dlp = new DlpServiceClient();
{{< /tab >}}
{{< tab header="Terraform" >}}

> Clients run `terraform init`.

Reviewer: Throw into a comment line

Reviewer: clarify to "Clients run `terraform init` to initialize Terraform in the current working directory."

Author: removed

    // IllegalArgumentException is thrown when an invalid argument has been passed to a function.
}
{{< /tab >}}
{{< tab header="Terraform" >}}

> Because they increase cyclomatic complexity, do not use error handling such as

Reviewer 2: Prioritizing between guidelines is something we avoid, instead prefer to cite the principles: are we saying this error handling technique is not in line with our simplicity goals?

We also prefer positive examples of what to do, rather than focusing on anti-patterns: what is the error handling we want samples to use? I expect it is mostly "none" because we want to rely on terraform to fall through to default logging, both for our maintenance and for developer practice.

suggestion: Start by recommending the error handling practice we want, then add special considerations around try.

#StyleGuideFeedback #TerraformPracticeFeedback

Author: done

## Testing {#testing}

### Cloud Shell

When possible, perform testing in Cloud Shell. Cloud Shell runs in a Compute
Engine virtual machine. The service credentials associated with this virtual
machine are automatic, so there is no need to set up or download a service
account key.

Cloud Shell provides you with command-line access to your cloud resources
directly from your browser.

[ide.cloud.google.com](ide.cloud.google.com) extends Cloud Shell with an
online development environment that includes:

* Cloud-native development via Cloud Code plugin support

* Rich language support for Go, Java, .Net, Python and NodeJS 

* Additional features such as integrated source control and support for
  multiple projects

### Test infrastructure

We recommend using a test infrastructure that spins up a clean project
for each test.

Testing infrastructure code means that you are deploying actual resources.
To avoid incurring charges, consider implementing a clean-up step.

> Some testing frameworks have a built-in cleanup step for you.

Author: Deleting this content for now and tracking discussion in b/271919695

  puts "Contents: #{content}"
end
{{< /highlight >}}
{{< /content-tab >}}
{{< content-tab header="Terraform" >}}
#### No CLIs

Do not include CLI commands (such as gcloud or kubectl) inside

> of your sample via the `null_resource` resource.

Author: Added that nuance

#### No Variables

Don't include Terraform variables (`var.VARIABLE_NAME`). Variables
enabling users to use Terraform samples not as snippets but as direct
module references. Disabling the avenue of customization 
reduces the risk so that users don't take accidental 
dependency.
Instead, hard code the resource names and attribute values.
In some cases, [locals](#locals) are allowed.

#### Locals {#locals}

> Sometimes multiple resources have a common string that is not
> accessible as a reference or there is no child/parent
> reference. Locals are allowed and encouraged for repeated text
> that a user may want to change. When a name is used multiple
> times, it's easier to copy, paste, and edit a local than search
> and replace the whole file. Locals are allowed and variable blocks
> are not because locals are an internal implementation and not
> part of the API  exposed via Terraform. Locals should be used only
> when necessary, and are best used when a repeated value is used
> throughout the configuration (3 or more times).

Reviewer 2: suggestion: We try to keep guidelines concise and actionable. Could we edit this down, reduce the amount of justification text, and add an example?

Here's an example of reducing a first paragraph to the core concept a developer needs to know:

```suggestion
Use local variables to reuse a common string 3 or more times that is not
accessible as a resource reference.
```

We could add a second paragraph with more insight if required.

#StyleGuideFeedback

Author: Done

#### Outputs

Output statement provide values that are generated after creating
a resource. They are allowed because they are readonly and they
give  users some important information such as URLs and IP addresses.
To keep it simple, they must be defined inline at the end of
a  sample, not in a separate `outputs.tf` file.

#### Resource references

> When possible, refer to already-defined resources.

Author: Added an example.

> #### Terraform Google Provider version

Reviewer 2: Provider version/argument/storage backend are things I mentioned above as maybe fitting into Client Initialization. Having them as broken out rules here seems fine too.

#StyleGuideFeedback

Author: Tracking in b/271024057, so closing here.

When your sample contains a new resource, add the required minimum
Terraform Google Provider version corresponding to which 
provider version introduced a resource. For example, `google_apigee_nat_address`
was added in version 4.37, so the minimum version is 4.37. To specify a minimum

> version in a sample, include the provider requirements, as follows:

Reviewer 2: This is a bit unexpected to me: I'm not sure this should be an exception from our general practice of not getting into "Why", but I'd like more insight into this personally. Where can I read more?

#TerraformPracticeFeedback

Author: Please see b/249338060 and https://github.com/terraform-google-modules/terraform-docs-samples/pull/177#issuecomment-1259969560

terraform {
  required_providers {
    google = {
      source = "hashicorp/google"
      version = "~= 4.37.0"
    }
  }
}

To determine when a resource was added, see the terraform-provider-google/CHANGELOG.md or 
terraform-provider-google-beta/CHANGELOG.md.

#### Provider argument

Add the provider argument if the resource is from provider google-beta.
resource "google_<resource_name>" "default" {
  provider = google-beta
  ...
}

Fields and resources that are only present in google-beta are clearly marked in
the provider documentation.
{{< /content-tab >}}

{{< /content-tabpane >}}

### Testing practices

> {{< content-tabpane >}}

Reviewer 2: blocker: We are in a section of the style guide where all the guideilnes are meant to be inside language-specific tab panes. For this testing practices heading to be excluded from Terraform and Ruby here means something is not organized correctly. Previewing would probably surface the problem quickly.

#StyleGuideFeedback

Author: fixed

{{< content-tab header="Terraform" >}}
#### Don't use `null_resource`

Do not include CLI commands (such as gcloud or kubectl) inside
of your sample via the `null_resource` resource.

# https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource

There are some products without Terraform support. Do not include
those products in Terraform samples.

#### No global variables

Don't include global Terraform variables (`var.VARIABLE_NAME`).

{{% alert title="Explanation" %}}
Global variables become part of the API exposed via Terraform.
Therefore, global variables help enable users to use Terraform samples
not as snippets but as direct module references. 
By disallowing global variables, we disable this avenue of customization.
This reduces the risk of users developing code that is
accidentally dependent on our samples.

Instead of including global variables, hard code the resource names and
parameter values.

Local variables are allowed in Terraform samples
because local variables are an internal implementation and therefore aren't
part of the API exposed via Terraform. See [local variables](#locals).
{{% /alert %}}

#### Local variables {#locals}

> Unlike global veriables, local variables are allowed in Terraform samples
> because local variables are an internal implementation and therefore aren't
> part of the API exposed via Terraform.
> 
> Use local variables to reuse a common string 3 or more times that is not
> accessible as a resource reference.

Reviewer 2: ```suggestion
Use local variables to reuse a common string 3 or more times that is not
accessible as a resource reference.

See [No Global Variables](#no-global-variables) for further explanation.
```

{{< /content-tab >}}

{{< content-tab header="Terraform" >}}
#### No CLIs

Do not include CLI commands (such as gcloud or kubectl) inside
of your sample via the `null_resource` resource.

# https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource

There are some products without Terraform support. Do not include
those products in Terraform samples.

#### No global variables

Don't include global Terraform variables (`var.VARIABLE_NAME`).

Global variables become part of the API exposed via Terraform.
Therefore, global variables help enable users to use Terraform samples
not as snippets but as direct module references. 
By disallowing global variables, we disable this avenue of customization.
This reduces the risk of users developing code that is
accidentally dependent on our samples.

Instead of including global variables, hard code the resource names and
parameter values.

In some cases, [local variables](#locals) are allowed.

#### Local variables {#locals}

Unlike global veriables, local variables are allowed in Terraform samples
because local variables are an internal implementation and therefore aren't

> part of the API exposed via Terraform.

Author: I added the wrapper, but this content was already in the global variables section, so I didn't move it anywhere.

Use local variables to reuse a common string 3 or more times that is not
accessible as a resource reference.

#### Resource references

> When possible, refer to already-defined resources. For example,

Author: Done

the `target` argument is referring to a `google_compute_target_ssl_proxy`
resource called `default` to get the `id` attribute of that resource.

```
resource "google_compute_global_forwarding_rule" "default" { 
  target = google_compute_target_ssl_proxy.default.id # Reference to already defined resource in the sample

  # Other esource arguments ...
}
```

#### Terraform Google Provider version

When your sample contains a new resource, add the required minimum
Terraform Google Provider version corresponding to which 
provider version introduced a resource. For example, `google_apigee_nat_address`
was added in version 4.37, so the minimum version is 4.37. To specify a minimum
version in a sample, include the provider requirements, as follows:

terraform {
  required_providers {
    google = {
      source = "hashicorp/google"
      version = "~= 4.37.0"
    }
  }
}

To determine when a resource was added, see the terraform-provider-google/CHANGELOG.md or 

> terraform-provider-google-beta/CHANGELOG.md.

Author: Done

[minitest]: https://github.com/minitest/minitest

{{< /content-tab >}}
{{< content-tab header="Terraform" >}}

> For information about testing, see the Terraform samples [contributing guide]

Reviewer: Thow into comment block. markdown hyperlink does not render in tabs

Author: removed

> (https://github.com/terraform-google-modules/terraform-docs-samples/blob/main/CONTRIBUTING.md).

Reviewer 2: blocker: We don't link to separate CONTRIBUTING guides from the style guide. However, after this PR is merged, the terraform-docs-samples repo should remove all redundant content and link here.

If there are some universal considerations on terraform sample testing that should existing beyond that repository, then we can add them here. Also, there is a testing section further up in the style guide that only lists Coverage currently, happy to discuss more guidelines there. We want to minimize the guidelines in the language-specific section at the bottom.

#StyleGuideFeedback

Author: Removed this and added two new sections to the general testing guidelines above. PTAL.
